### PR TITLE
week 3 notes

### PR DESCRIPTION
**Pull request by ajtowns**

### CODE REVIEW

File: week-3.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+One thing that's been working out really well (in my opinion, anyway!) is how engaged reviewers have been with the regular process of filing issues and pull-requests on github to make improvements to the drafts. Since that's the usual way of improving most open source Bitcoin related things, we've been encouraging people to focus on that, rather than collecting review results via google forms or similar. But to help avoid things getting missed, we've added some free form fields to the knowledge check survey for capturing things that might need changes that haven't been filed in github.
+
+## Week 1 Feedback
```
Comment (bitschmidty) [inline week-3.md:13]: Standardize on Week 1 vs Week one?

File: week-3.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+This poll was just for the first week, so it's already out of date, but since this is all an experiment, might be of interest anyway.
+
+## Results from Week Two
+
+We've had the second two Q&A sessions, with meetbot logs: [week 2 Q&A 1](), [week 2 Q&A 2]() -- but as it turns out the second Q&A didn't actually have any questions. Since Q&A 2 in week one seemed a bit sparser too, we'll just advertise the single slot this week, but if that's not convenient for you, do feel free to ask questions anytime -- experts are lurking in that channel most of the time, and even if they're not available to answer immediately, if you can stay connected you will often get an answer sometime later. (If you can't stay connected, you can use the [logs](http://www.erisian.com.au/taproot-bip-review/) to catch your answer, but it might be wise to mention that you'll be doing so to avoid people thinking there's no point responding because you've left the channel)
```
Comment (bitschmidty) [inline week-3.md]: http://www.erisian.com.au/meetbot/taproot-bip-review/2019/taproot-bip-review.2019-11-12-19.00.log.html
and 
http://www.erisian.com.au/meetbot/taproot-bip-review/2019/taproot-bip-review.2019-11-14-02.00.log.html

File: week-3.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+> "that may have been too obvious for me when writing it" -- Pieter Wuille, [personal correspondence 2019-11-13](#correspondence)
+
+One thing that might not be obvious is that after working on something complicated and technical like this, it's possible to forget which things are obvious to everyone and which things are only obvious after you've been studying the subject matter deeply. The quote above is in regard to the bech32 issue discussed last week and that, and how something that seems like a surprising bug to some might just be a completely expected trade off to someone more deeply steeped in the technology. Hopefully wide review amongst people with varying levels of knowledge and experience can help reduce these sorts of misunderstandings.
```
Comment (bitschmidty) [inline week-3.md]: Is this an appropriate link for "bech32 issue discussed" text?
http://www.erisian.com.au/taproot-bip-review/log-2019-11-05.html#l-231

File: week-3.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+You may wish to skim the "Applications" section since we've got week 5 reserved for looking into that in more detail. However most of the complexity there is actually in reading the detailed explanations of how those work in other papers and documents, so just reading the whole BIP this week might be a good approach too.
+
+Jonas Nick wrote a [blog post](https://medium.com/blockstream/reducing-bitcoin-transaction-sizes-with-x-only-pubkeys-f86476af05d7) for the Blockstream blog about x-only pubkey construction that might be interesting. t-bast from group 17 pointed out a [paper on bad nonces with ECDSA](https://eprint.iacr.org/2019/023.pdf) that may be useful background for why deterministic nonce generation is useful when possible; and Leonid Beder from the same group pointed to a [talk by Andrew Poelstra at the Monero Conference on Schnorr signatures](https://www.youtube.com/watch?v=L6KqkrP_nU4) that mentions an interesting construction to ensure proper nonce generation by hardware wallets. The videos from the [optech taproot workshop](https://bitcoinops.org/en/schorr-taproot-workshop/) might be helpful as a slightly gentler introduction to the topic.
```
Comment (bitschmidty) [inline week-3.md:44]: Perhaps this optech workshop link instead:
https://bitcoinops.org/en/schorr-taproot-workshop/#schnorr-signatures-and-musig

File: week-3.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+If you're interested in looking at implementations, then you may wish to look at any or all of:
+
+ * the [python reference code](https://github.com/sipa/bips/tree/bip-schnorr/bip-schnorr/) (not for production use)
+ * the [jupyter notebooks from the optech taproot workshop](https://bitcoinops.org/en/schorr-taproot-workshop/) (also not for production use)
```
Comment (bitschmidty) [inline week-3.md]: Could alternatively use https://github.com/bitcoinops/taproot-workshop/

**Review by bitschmidty — COMMENTED** [review]
@ajtowns thanks for putting this together, looking great. Left a couple comments.

Also pushed a commit for adding some of the week links, fc0e9f374a9efbb378fe1e447155a891539741c3

**Comment by harding** [comment]
LGTM, assuming at least the missing links get filled in with @bitschmidty's suggestions.  Nice work!

File: week-3.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ * Tue 19 Nov 20:00 +0100 Paris
+ * Wed 20 Nov 00:30 +0530 Calcutta
+ * Wed 20 Nov 04:00 +0900 Tokyo
+ * Wed 20 Nov 06:00 +1100 Sydney
+
+There's also the dedicated [slack instance](https://bitcoin-review.slack.com/) if you prefer that method of communicating.
+
+## Correspondence
+
+IRC conversation between Pieter Wuille (sipa) and David Harding (harding). Quoted with permission, links added and some minor typographical errors corrected.
```
Comment (sipa) [inline week-3.md]: About including this: I don't object to publishing it, but I feel that perhaps it is a distraction.

There are two issues here. The first is that Bech32 was too vague about only having guarantees against substitution errors (it talks about "number of wrong characters" and "error affecting N characters", which can probably be interpreted either way). This was a problem with the writeup. But if people assumed that Bech32 did in fact have guarantees against insertion/deletion errors, there would still not really be a problem. They'd be wrong about thinking the chance of missing detection for those is 0, while it is in fact 1 in a billion - but it would be unlikely for another design to offer better detection properties without worsening overall properties.

The other issue is that regardless of the specific language used in BIP173, a very reasonable expectation was broken. Any type of error (including ones without strict guarantees) should still be detected with very high probability (BIP173 states things like "As the number of errors goes up, the chance converges towards 1 in 2^30"). This specific insertion/deletion error violates that: it has a 1/32 chance of being not detected, and could have been avoided by better review or analysis by someone more familiar with error detection codes.

I feel that including the conversion moves the focus from the second (a serious design mistake) to the first (being vague about a minor and unavoidable issue).

Comment (harding) [reply week-3.md]: @sipa

> Bech32 was too vague about only having guarantees against substitution errors (it talks about "number of wrong characters" and "error affecting N characters", which can probably be interpreted either way).

Just to be clear, I think better context for that second quote from BIP173 is "**any** error affecting N characters" (emphasis different than original).  As explained in the conversation quoted in this PR, I think that "any error" comes across as something that would include insertion/deletion errors.  Additionally, I believe the "any error" line is the first place where error-detection probability is discussed in the document, so it sets the expectation for readers that other parts of the document that mention errors are describing *any* error.

> if people assumed that Bech32 did in fact have guarantees against insertion/deletion errors, there would still not really be a problem. They'd be wrong about thinking the chance of missing detection for those is 0, while it is in fact 1 in a billion

I don't understand your point here.  In my mind, the problem with people operating under false assumptions when reviewing a specification is only partly that they may receive a different result than expected.  It's also that their understanding is incomplete and so they don't have the comprehension necessary to ask good follow up questions.

Perhaps if I had known that there was a small chance of missing the detection of ins/del errors, I would've asked for more information.  For example, even now knowing about this problem, I'm not sure how it affects the probability of detecting up to four errors of *any* type.  E.g., does it only fail in 1-in-a-billion times for any changed length, or is it 4-in-a-billion if I make four length changes, or is it something else?  Maybe I would've asked for a variation on @gmaxwell's chart showing the detection probability of bech32 vs base58check, since it seems to me that bech32 having at least a 1-in-a-billion chance of failing due to insertion/deletion errors means there's no way it can perform better than base58check.

 

(Chart sourced from [here](https://www.reddit.com/r/Bitcoin/comments/9b62hb/bech32_address_concerns/e50xt5i/), where Maxwell says, "so long as you make errors like a human does-- a few errors rather than just totally random nonsense that happens to match the character set and length-- the 30-bit checksum in Bech32 is much more protective than a longer 32-bit cryptographic hash."   In retrospect, it seems obvious to me that humans will make insertion/deletion errors.)

> I feel that including the conversion moves the focus from the second (a serious design mistake) to the first (being vague about a minor and unavoidable issue).

I think we would've been much more likely to have caught the "serious design mistake" if more people were aware of the "minor and unavoidable issue".  I think that's relevant to the review club because many reviewers (such as myself!) lack the background and experience to catch serious design mistakes, but most reviewers can look for parts of the specification that are "vague" and ask for clarification, which can help experts discover where there might be undiscovered serious issues.

I think the goal of including the quote in this document is to help encourage reviewers by showing them how even just asking for details about vague or ambiguous wording in the specification may ultimately lead to important new insights.  I think that's good, but I'm also fine with the quote being removed if you (or anyone else) still thinks that it's inappropriate.

Comment (sipa) [reply week-3.md]: > Just to be clear, I think better context for that second quote from BIP173 is "any error affecting N characters" (emphasis different than original). As explained in the conversation quoted in this PR, I think that "any error" comes across as something that would include insertion/deletion errors. Additionally, I believe the "any error" line is the first place where error-detection probability is discussed in the document, so it sets the expectation for readers that other parts of the document that mention errors are describing any error.

Yes, I agree this is confusing, and should be improved. I'm not dismissing the fact that the document is vague; I'm arguing it shouldn't be conflated with the fact that there was a distinct and much more serious problem.

I believe that anyone familiar with error correcting codes would have correctly assumed that "an error affecting N characters" implied substitution errors - insertions and deletions move positions so will naturally affect almost every character.

> For example, even now knowing about this problem, I'm not sure how it affects the probability of detecting up to four errors of any type. E.g., does it only fail in 1-in-a-billion times for any changed length, or is it 4-in-a-billion if I make four length changes, or is it something else?

Bech32 optimizes for substitution errors, up to 4 of which are guaranteed to be detected. For any other class of errors - it doesn't matter where or how or how many, the goal was (approximately) 1 in a billion chance of not being detected. I believe that is the case as well, except for "inserting/deleting a q in the second to last position", which apparently only has a 1/32 chance.

> I think we would've been much more likely to have caught the "serious design mistake" if more people were aware of the "minor and unavoidable issue".

I'm not convinced about that - perhaps even the opposite. The problem was found by someone who was writing tests for Bech32, and including insertion/deletion errors in the mutations he was trying, perhaps exactly because he believed that the spec claims a guarantee would detect them.

File: week-3.md (hunk 4 above)
Comment (ajtowns) [reply week-3.md:44]: I dropped the internal link, because I thought the elliptic curve math stuff might be useful to some, and anyone for whom it's not can easily skip over it

File: week-3.md (hunk 5 above)
Comment (ajtowns) [reply week-3.md]: er, yeah, i think i mispasted the url there

File: week-3.md (hunk 6 above)
Comment (ajtowns) [reply week-3.md]: I still think there's something useful here to raise, but I'm going to skip trying to include it this week

File: week-3.md (hunk 3 above)
Comment (ajtowns) [reply week-3.md]: dropped this paragraph

File: week-3.md (hunk 6 above)
Comment (harding) [reply week-3.md]: I think @ajtowns is making the right decision by yanking the section from this week's newsletter.  To address @sipa's reply:

> I'm not convinced about that - perhaps even the opposite. The problem was found by someone who was writing tests for Bech32, and including insertion/deletion errors in the mutations he was trying, perhaps exactly because he believed that the spec claims a guarantee would detect them.

I agree, it seems quite possible that person wouldn't have written that test, or at least not have investigated further after the first failure, if the test failure didn't break an inferred guarantee.

OTOH, I offer you an alternative:  what if someone had discovered the specification was unclear about different types of errors and requested it be clarified---which I'm sure you would've done.  Then, I think it's possible that person or someone else would've suggested including the witness program length in the data covered by the bech32 checksum.  (This seems to me like a reasonable theory given that ZmnSCPxj [suggested](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-November/017448.html) exactly that less than seven hours after he perhaps first learned about the problem from your post,.)  IIUC, that would bump up the reliability of variable-length v1+ bech32 addresses to the "any error" guarantee BIP173 currently implies.  If that change had been proposed while bech32 was still under review and not widely adopted, I think it would've been an easy change to support.

In short, if the "minor and unavoidable issue" had been described explicitly, perhaps as the result of a reviewer being confused by BIP173's claims, I think we might've avoided the "serious design mistake" altogether even without learning about it.  If true, I again think that's useful motivational material to convey to reviewers who are often afraid that they won't be able to make valuable contributions because they aren't as knowledgeable as the original proposers.  Just clarifying ambiguities in BIPs has the guaranteed result of making the document more informative, and a more informative document may reveal problems that would otherwise have gone unnoticed and unaddressed.

Comment (sipa) [reply week-3.md]: > Then, I think it's possible that person or someone else would've suggested including the witness program length in the data covered by the bech32 checksum.

We're far into hypotheticals here, but I feel you really misunderstand the issue here.

Ignoring the insert-q-in-penultimate-position issue for a minute, *anything* but up to 4 substitution errors has a 1 in a billion failure to detect chance. Adding the length into the checksummed data would not change this - an insertion or deletion error would still have a 1 in a billion failure to detect chance. Including the length does not magically turn insertions or deletions into substitution errors.

So I think if anyone would have suggested including the length explicitly without having the understanding of the insert-q bug, it would have been dismissed as needless complication.

Yes, adding the length into the checksum is a workaround for the 1/32 failure chance - but there are simpler solutions too (see my last mail to the list).

I of course agree that being clearer in the BIP on the exact guarantees is better, and I'll try to improve it now. But to me it really feels that this unclarity is an unrelated issue.

Comment (harding) [reply week-3.md]: > I feel you really misunderstand the issue here.

That's quite likely.  I'll stop commenting.

File: week-3.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+One thing that's been working out really well (in my opinion, anyway!) is how engaged reviewers have been with the regular processes of communicating online (irc, bitcoin-dev, etc), and filing issues and pull-requests on github to make improvements to the drafts. Since that's the usual way of improving most open source Bitcoin related things, we've been encouraging people to focus on that, rather than collecting review results via google forms or similar. But to help avoid things getting missed, we've added some free form fields to the knowledge check survey for capturing things that might need changes that haven't been filed in github.
```
Comment (jonatack) [inline week-3.md]: suggestions:

s/pull-requests on github to make improvements to/pull requests on GitHub to improve/

s/Bitcoin related/Bitcoin-related/

s/free form/free-form/

s/github/GitHub/

File: week-3.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+With luck, everyone's into the swing of things now, because we're moving on to the first look at the final BIP, both the alpha and omega, if you will: [bip-schnorr](https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki). This marks the halfway point -- by the end of this week we'll have gone through all the BIPs (if you're following the [curriculum](Curriculum.md) anyway!) and for the following three weeks we'll be focussing on some of the details (that you might have skipped over if you followed the suggestions here) and their implications.
+
+The quote this week emphasises the widely known wisdom that you shouldn't roll your own crypto -- even though that's essentially what we're doing here with both a new standard for schnorr signatures, and introducing the taproot construction for real world use. Both of these are security constructions that rely on mathematical assumptions working out as expected; so before adopting it, it's wise to make sure that these new constructions really can survive expert cryptanalysis. (We'll see in the "schnorr applications" week how the first version of the musig construction didn't survive expert review, eg)
```
Comment (jonatack) [inline week-3.md:9]: nit: missing period at end of sentence and could probably remove ", eg" or replace it with s/We'll/For instance, we'll/

File: week-3.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+## Week 1 Feedback
+
+We got 36 responses (about 20%) to the "week 1 feedback" survey -- summary stats were:
+
+ * Solo review: 2 hours (15 responses), 3+ hours (11 responses), 1 hour (6), 30 min (4)
+ * Study group: 1 hour (13), 2 hours (8), None! (8), 30 minutes (6), 3+ hours (1)
+ * Q&A session: 1 hour (19), 30 minutes (7), None! (5), 2 hours (4), 3+ hours (1)
+ * PoC work: None! (33), 3+ hours (1), 1 hour (1), 30 minutes (1)
+
+Most people rated their first week progress as 2 or 3, so about as expected, and resources as 3-5 which is great. Study group rating was bit more spread out, with nine responses give a low rating of 1 or 2, twenty giving mid-good rating of 3 or 4, and six giving top marks of 5. So if your group is doing particularly well, it might be worth writing a blog post or making a youtube video or similar about what's going right so others can learn from you; alternatively if your group is not doing so great, it might be interesting to lurk through some of the other groups' slack channels to see what works for them and perhaps try it in your group.
```
Comment (jonatack) [inline week-3.md]: suggestions:

s/give a low/giving a low/, s/mid-good/a mid-good/, s/giving top marks/giving the top mark/
or
s/give a low rating/giving low ratings/, s/mid-good rating/mid-good ratings/

s/lurk through/lurk in/

File: week-3.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+This poll was just for the first week, so it's already out of date, but since this is all an experiment, might be of interest anyway.
+
+## Results from Week Two
+
+We've had the second two Q&A sessions, with meetbot logs: [week 2 Q&A 1](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/taproot-bip-review.2019-11-12-19.00.log.html), [week 2 Q&A 2](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/taproot-bip-review.2019-11-14-02.00.log.html) -- but as it turns out the second Q&A didn't actually have any questions. Since Q&A 2 in week one seemed a bit sparser too, we'll just advertise the single slot this week, but if that's not convenient for you, do feel free to ask questions anytime -- experts are lurking in that channel most of the time, and even if they're not available to answer immediately, if you can stay connected you will often get an answer sometime later. (If you can't stay connected, you can use the [logs](http://www.erisian.com.au/taproot-bip-review/) to catch answers to your question, but it might be wise to mention that you'll be doing so to avoid people thinking there's no point responding because you've left the channel)
```
Comment (jonatack) [inline week-3.md:28]: nit: missing period at end of sentence

File: week-3.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+There's been a few more suggested improvements for the BIPs, including [#138](https://github.com/sipa/bips/issues/138), [PR#145](https://github.com/sipa/bips/pull/145), [PR#144](https://github.com/sipa/bips/pull/144), [PR#137](https://github.com/sipa/bips/pull/137), and [PR#136](https://github.com/sipa/bips/pull/136). Max Hillebrand has filed an [issue for high-level discussion of adding taproot to Wasabi Wallet](https://github.com/zkSNACKs/WalletWasabi/issues/2531) that might be interesting to see.
```
Comment (jonatack) [inline week-3.md:30]: nit: s/There's/There have/

There have been quite a few, so it's unclear why these PRs are mentioned (and they have been merged since). Also, the issue mentioned has seen almost no discussion. This paragraph could probably be dropped or provide just a link to sipa/bips/pulls.

File: week-3.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+If you're interested in looking at implementations, then you may wish to look at any or all of:
+
+ * the [python reference code](https://github.com/sipa/bips/tree/bip-schnorr/bip-schnorr/) (not for production use)
+ * the [jupyter notebooks from the optech taproot workshop](https://github.com/bitcoinops/taproot-workshop/) (also not for production use)
+ * [libsecp256k1 PR#558](https://github.com/bitcoin-core/secp256k1/pull/558) adds basic support schnorr signature
```
Comment (jonatack) [inline week-3.md]: s/support schnorr signature/schnorr signature support/

File: week-3.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+ 
+## Misc
+
+The IRC channel for the Q&A sessions is ##taproot-bip-review on the Freenode IRC network. If you've got questions through the week, feel free to ask there -- if someone is around and available they might be able to answer. As well as the [meeting logs](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/) there are ongoing [channel logs](http://www.erisian.com.au/meetbot/taproot-bip-review/2019/) as well, in case you want to see what's been going on there while you've been absent. As mentioned above, there's only one scheduled Q&A session this week: Tue 1900 UTC (America/Europe/Africa).
```
Comment (jonatack) [inline week-3.md]: Same url is provided in both links.

**Review by jonatack — COMMENTED** [review]
Great stuff. A few comments and nits below; feel free to ignore.

File: week-3.md (hunk 7 above)
Comment (moneyball) [inline week-3.md]: We haven't sent out the knowledge check survey yet, so I don't think anyone will know what it is. When we do, we should link to it so people have a reference.

Comment (moneyball) [inline week-3.md]: Should probably be "in our opinion" since later in the email "we" is used (vs. "I"). Even though this is all written by @ajtowns.

Comment (jonatack) [reply week-3.md]: I didn't catch that this was different survey from the first one, but it sounds like an excellent idea. If it's too early to go into more detail, perhaps simply: s/knowledge check survey/upcoming surveys/

Comment (jonatack) [reply week-3.md]: Could simplify/shorten with: "One thing that's been working out well is..."

File: week-3.md (hunk 12 above)
Comment (ajtowns) [reply week-3.md]: changed

File: week-3.md (hunk 13 above)
Comment (ajtowns) [reply week-3.md]: probably a bug in week-2 as well then
